parquet_derive/src/lib.rs

```diff
@@ -0,0 +1,294 @@
+      o => unimplemented!("don't know {} for {:#?}", o, f),
+    };
+
+    FieldInfo {
+      field_name: f.ident.clone().expect("must be a named field"),
+      field_lifetime,
+      field_type,
+      is_option,
+      column_writer_variant,
+    }
+  }
+
+  pub fn to_writer_snippet(&self) -> proc_macro2::TokenStream {
+    let field_name = self.field_name.clone();
+    let field_type = self.field_type.clone();
+    let column_writer_variant = self.column_writer_variant.clone();
+    let is_option = self.is_option;
+
+    if is_option {
+      match &self.field_type.to_string()[..] {
+        "str" => quote! {
+            {
+                let definition_levels : Vec<i16> = self.iter().
+                        map(|x| x.#field_name).
+                        map(|y| if y.is_some() { 1 } else { 0 }).
+                        collect();
+                let vals : Vec<parquet::data_type::ByteArray> = self.iter().
+                        map(|x| x.#field_name).
+                        filter_map(|z| {
+                          if let Some(ref inner) = z {
+                              Some((*inner).into())
+                          } else {
+                              None
+                          }
+                        }).
+                        collect();
+                if let #column_writer_variant(ref mut typed) = column_writer {
+                    typed.write_batch(&vals[..], Some(&definition_levels[..]), None).unwrap();
+                }
+            }
+        },
+        // TODO: can this be lumped with str by doing Borrow<str>/AsRef<str> in the
+        // ByteArray::from?
+        "String" => {
+          quote! {
+            {
+                let definition_levels : Vec<i16> = self.iter().
+                        map(|x| &x.#field_name).
+                        map(|y| if y.is_some() { 1 } else { 0 }).
+                        collect();
+                let vals : Vec<parquet::data_type::ByteArray> = self.iter().
+                        map(|x| &x.#field_name).
+                        filter_map(|z| {
+                          if let Some(ref inner) = z {
+                              Some((&inner[..]).into())
+                          } else {
+                              None
+                          }
+                        }).
+                        collect();
+                if let #column_writer_variant(ref mut typed) = column_writer {
+                    typed.write_batch(&vals[..], Some(&definition_levels[..]), None).unwrap();
+                }
+            }
+          }
+        },
+        _ => quote! {
+            {
+                let filtered_vals : Vec<#field_type> = self.iter().
+                        map(|x| x.#field_name).
+                        filter(|y| y.is_some()).
+                        filter_map(|x| x).
+                        collect();
+                let definition_levels : Vec<i16> = self.iter().
```

Comment:
Definition levels are collected in the same way. Can we factor it out and do it once for all of them?

Reply:
Done!